apps/extension/src/routes/popup/popup-needs.ts

```diff
@@ -0,0 +1,18 @@
+import { redirect } from 'react-router-dom';
+import { PopupPath } from './paths';
+import { localExtStorage, sessionExtStorage } from '@penumbra-zone/storage';
+
+export const needsLogin = async (): Promise<Response | undefined> => {
+  const password = await sessionExtStorage.get('passwordKey');
+  if (password) return;
+
+  return redirect(PopupPath.LOGIN);
+};
+
+export const needsOnboard = async () => {
+  const wallets = await localExtStorage.get('wallets');
+  if (wallets.length) return;
+
+  void chrome.runtime.openOptionsPage();
+  window.close();
+};
```

Comment:
These need to return null right? Otherwise there will be that extension update error we saw while pairing.